Sticky Position: Try re-enabling non-root sticky position

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of #47043, will ideally resolve https://github.com/WordPress/gutenberg/issues/47892

This PR explores re-enabling sticky position at non-root positions within the document, and adds paragraph / help text beneath the position drop-down selection when Sticky is set, so that we describe in text that the block will be sticky to its parent.

Early versions of this PR also explored adding in a position visualizer of some kind, however it was uncertain as to how to make this truly beneficial or useful. For now, this PR has been pared back to just re-enabling non-root positions, and adding in the help text.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

For the initial version of sticky positioning, support for sticky being set at non-root positions was removed because of the potential confusion for users. The goal with this PR is to explore what sorts of things we can do in the UI to help ease some of that confusion, and to determine whether or not it's enough to re-enable non-root position values.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

<!--
* Add a position visualizer, borrowing the approaches from the padding and margin visualizers, and pass it the clientId of the parent of the current block, and only display when `sticky` is set. -->
* Enable sticky position at non-root positions.
* Add help text beneath the drop-down select when `sticky` is set.

## To-do items

* [x] ~Resizing the browser doesn't update the dimensions of the popover~
* [x] T~he block popover extends beyond the screen and creates an additional scrollbar (Fix in: https://github.com/WordPress/gutenberg/pull/49978)~
* [x] ~If you hover over the parent block and attempt to scroll, it doesn't let you~
* [x] ~The border of the block popover renders over the top of the block toolbar~
* [x] ~Determine when to show / hide the popover (shows while block is selected for now)~
* [x] ~The size calculation doesn't quite work in non-100% browser zoom levels~

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Copy / paste some test markup below (a sticky group in a column):

<details>

<summary>Block markup for a sticky group within a column in a columns block</summary>

```html
<!-- wp:columns {"align":"wide","style":{"spacing":{"padding":{"top":"var:preset|spacing|30","right":"var:preset|spacing|30","bottom":"var:preset|spacing|30","left":"var:preset|spacing|30"}}},"backgroundColor":"tertiary"} -->
<div class="wp-block-columns alignwide has-tertiary-background-color has-background" style="padding-top:var(--wp--preset--spacing--30);padding-right:var(--wp--preset--spacing--30);padding-bottom:var(--wp--preset--spacing--30);padding-left:var(--wp--preset--spacing--30)"><!-- wp:column {"width":"40%"} -->
<div class="wp-block-column" style="flex-basis:40%"><!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|30","right":"var:preset|spacing|30","bottom":"var:preset|spacing|30","left":"var:preset|spacing|30"}},"position":{"type":"sticky","top":"0px"}},"backgroundColor":"cyan-bluish-gray","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-cyan-bluish-gray-background-color has-background" style="padding-top:var(--wp--preset--spacing--30);padding-right:var(--wp--preset--spacing--30);padding-bottom:var(--wp--preset--spacing--30);padding-left:var(--wp--preset--spacing--30)"><!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading">A sticky card</h3>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>Here is a description to be contained within this sticky element within the sidebar. It could be a little information about the author, or it could include a button of some kind.</p>
<!-- /wp:paragraph -->

<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"backgroundColor":"contrast","textColor":"base"} -->
<div class="wp-block-button"><a class="wp-block-button__link has-base-color has-contrast-background-color has-text-color has-background wp-element-button">Contact</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons --></div>
<!-- /wp:group --></div>
<!-- /wp:column -->

<!-- wp:column {"width":"60%"} -->
<div class="wp-block-column" style="flex-basis:60%"><!-- wp:group {"layout":{"type":"flex","flexWrap":"wrap","justifyContent":"right"}} -->
<div class="wp-block-group"><!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"pale-cyan-blue","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-pale-cyan-blue-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"light-green-cyan","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-light-green-cyan-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"light-green-cyan","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-light-green-cyan-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"pale-cyan-blue","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-pale-cyan-blue-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"pale-cyan-blue","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-pale-cyan-blue-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"light-green-cyan","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-light-green-cyan-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"light-green-cyan","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-light-green-cyan-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"layout":{"selfStretch":"fixed","flexSize":"300px"},"dimensions":{"minHeight":"250px"}},"backgroundColor":"pale-cyan-blue","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-pale-cyan-blue-background-color has-background" style="min-height:250px"></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

</details>

Or, construct your own pattern:

1. Add a Group block as a child of a Group block that also contains some paragraph text.
2. Add a background color to the parent Group block so that it's a little clearer what's going on.
3. Add paragraph text after the parent Group block so that you can see the scroll behaviour take effect.
4. In the nested Group block, give it its own background color, and add some paragraph text to it.
5. Set this nested Group block to "sticky" and test the UI.

## Screenshots or screencast <!-- if applicable -->

<img width="1276" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/5b59dfb9-d392-4ba2-8da5-a305cab06521">

<!--

![2023-05-26 14 29 32](https://github.com/WordPress/gutenberg/assets/14988353/ff4fffa0-37a8-4523-9183-a454e1b405c7)

### Earlier work in progress screenshots and screengrabs of progress:

![image](https://user-images.githubusercontent.com/14988353/234486384-9ec5d80a-c0f1-46d3-977f-ac1b2d6582c1.png)

<img width="283" alt="image" src="https://user-images.githubusercontent.com/14988353/234486435-9f550821-0542-4d21-a648-eaf215916f8f.png">

Screengrabs:

Post editor:

https://user-images.githubusercontent.com/14988353/234491629-4b96f91a-1d0a-4e68-8ab3-1b8ad7f79fbb.mp4

-->

Site frontend:

https://user-images.githubusercontent.com/14988353/234487118-979634ab-116b-4815-aafb-08099e408ac7.mp4

andrewserong committed:
Sticky Position: Try re-enabling non-root sticky position, and add a visualizer to define sticky area

andrewserong committed:
Fix whitespace

andrewserong commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -203,6 +209,71 @@ export function useIsPositionDisabled( { name: blockName } = {} ) {
+	}
+
+	return (
+		<BlockPopover
```
Just a note: To fix the issue where the popover goes off the edge of the screen, it might be worth experimenting with `Popover` instead and then pass it a rect (or `getBoundingClientRect()` function) so that we can ensure it's only rendering over the visible area of the page. I.e. treat this a little bit like the drop indicators.

andrewserong replied:
Potential fix for this issue in https://github.com/WordPress/gutenberg/pull/49978

andrewserong committed:
Ensure the position visualizer does not interfere with the ability to select inner blocks

andrewserong commented:
Update:

* I've fixed up the issue where the popover prevented interacting with inner blocks (borrowing from the approach used in https://github.com/WordPress/gutenberg/pull/41153)
* Updated the `z-index` value so that the popover appears underneath the toolbar
* Opened up a fix for the additional scrollbars issue: https://github.com/WordPress/gutenberg/pull/49978

andrewserong committed:
Ensure resizing the block editor will update the block popover dimensions

andrewserong commented:
In https://github.com/WordPress/gutenberg/pull/49321/commits/f2f4682b68b1593bf3e750e657dd1e09310df1b2, I've fixed up the dimensions of the BlockPopover so that when the browser window is resized (or when scrollbars appear) then it gets recalculated. If it turns out that this shouldn't be behaviour for all BlockPopovers, then we can always move the logic into the PositionVisualizer instead. For now, I'll leave it where it is, and we can look at it in more detail once this PR is ready for review.

Edit: it could possibly be better to use a `useResizeObserver` instead, but I believe that would need to be output around the element that's being targeted, which we don't really have access to in React in the current context... but depending on feedback, it could be worth attempting to refactor to use that instead.

andrewserong commented:
This is getting closer to a testable state, but one issue I ran into is that the `top` position isn't aware of any other sticky elements on the page. So, for a site that has a sticky header and also wants other areas to be sticky, it doesn't automatically detect a correct `top` value to use that would be pleasing, and the non-root-sticky element winds up going above the sticky header:

https://user-images.githubusercontent.com/14988353/234474732-cede1b93-41b9-4a9f-91ce-5eb9988acaba.mp4

In a long post that contains sticky headings, this behaviour of sitting on top of the other sticky element could actually be desirable.

What I'm wonder now, is whether we might want to give the user the ability to set a top offset value for the sticky position?

andrewserong committed:
Try simplifying the display logic so it always shows the visualizer if the block is selected

andrewserong commented:
Update: this PR should be ready for review now — it's an experimental idea to see if adding text, plus a visualizer for sticky positioning will help improve the UX for sticky positioning at non-root positions.

Feedback and alternate ideas are very welcome! For example, we could do a lot of adjustment to how the sticky visualizer appears, tweak text, etc.

I noticed one potential block, which is that if the site uses sticky position for its header, then the element that's set to sticky doesn't know that it should have a `top` value to steer clear of it. A question for reviewers is whether or not re-enabling in non-root positions will necessitate having a UI control over the top position or offset.

jasmussen commented:
Thanks for considering this. It's an excellent instinct, the use case for non-root locations is valid, the video is helpful, and the example code even moreso. Here's a GIF just showing me experiencing the same as your demo:

![testing](https://user-images.githubusercontent.com/1204802/234501336-319368c5-bc2d-4094-a25d-e75977968566.gif)

The key take-away here is shown in this image:

<img width="1392" alt="Screenshot 2023-04-26 at 09 25 35" src="https://user-images.githubusercontent.com/1204802/234501374-0d772a55-8e5a-4995-9c58-d8a14ebd9437.png">

The light blue border shown in the canvas here is meant to denote the _parent container_, so that it's clear what the block sticks to.

This actually reminds me a fair bit of some similar explorations we've had in the past, in general around template parts, ancestors, etc, having explored dotted lines or similar. #44774 and #44775 are tangentially related. Because these block outlines have come up many times in the past, I'd appreciate input from @jameskoster or @SaxonF, because when to show outlines and what they should look like is something we need to get right. 

Is contextuality like suggested here the key? Should the outline be shown when the Sticky option is focused or interacted with, and ten fade out? Or should it show when the block is selected? Should it be shown when a block is hovered? Should it be a dotted line? Should it be something else, like dimming non-selected blocks, more like spotlight mode?

jameskoster commented:
I'm curious about the motivation to indicate the sticky block container on the canvas? It will always be the direct ancestor, so it doesn't seem super useful to me. I am likely missing something though.

Overall I wonder if it would be better expressed in List View. I think we tried some explorations around that before but didn't land anything yet.

andrewserong commented:
Thanks for the feedback and suggestions @jasmussen and @jameskoster!

> I'm curious about the motivation to indicate the sticky block container on the canvas? It will always be the direct ancestor, so it doesn't seem super useful to me. I am likely missing something though.

Good question — when we first landed the sticky position support, based on user feedback, it turned out that it was confusing for folks that when they set a non-root Group block to Sticky, they thought it'd stick to the top of the page, and the behaviour that it sticks to the direct ancestor wasn't clear. Based on that feedback, we removed the non-root sticky support until the UX could be improved (here's [a wrap up comment](https://github.com/WordPress/gutenberg/issues/47892#issuecomment-1423668583) from that time).

So, the problem to solve is how do we make it clear to a user what's happening when they set a block to `sticky`? My current hunch is that the best bet is to provide a visual clue within the editor canvas so that a user can see the visual area that defines how sticky will work. Highlighting the area of the parent in relation to the currently selected block seems like a good way to do it to me, but I'm sure there'd be other options!

> Overall I wonder if it would be better expressed in List View. I think we tried some explorations around that before but didn't land anything yet.

We haven't tried anything that flags which parent it relates to yet, I don't think. There's an open PR (https://github.com/WordPress/gutenberg/pull/49122) to flag the _status_ of sticky, but not the relationship between the sticky block and its parent. While improving things in the list view is a good idea, my feeling is that we'll need something a little more prominent in the editor canvas in order for the feature not to feel broken for users that set sticky in nested blocks but don't really know what's going on.

> Is contextuality like suggested here the key? Should the outline be shown when the Sticky option is focused or interacted with, and ten fade out? Or should it show when the block is selected? Should it be shown when a block is hovered? Should it be a dotted line? Should it be something else, like dimming non-selected blocks, more like spotlight mode?

These are exactly the kinds of questions I was hoping for! I think showing the context for the user is key, similar to how we have padding and margin visualizers to help users understand those concepts.

Determining exactly when to show the visualizer, when to hide it, and what it should look like are all great things to explore. Let me know if anyone has ideas about how this might look, and I'm happy to continue hacking around 🙂

ramonjd commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -222,32 +279,39 @@ export function PositionPanel( props ) {
+
 	const options = useMemo( () => {
 		const availableOptions = [ DEFAULT_OPTION ];
 		// Only display sticky option if the block has no parents (is at the root of the document),
```
Is this comment still valid? It looks like the parents test has been removed?

andrewserong replied:
Oh, it's not! I'll remove it 👍

ramonjd commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -222,32 +279,39 @@ export function PositionPanel( props ) {
 
+	const blockInformation = useBlockDisplayInformation( firstParentClientId );
+	const stickyHelpText =
+		allowSticky && value === 'sticky' && blockInformation
```
Would it be worth checking against `STICKY_OPTION.value`?

andrewserong replied:
That'd be neater, yes — I'll update. Thanks for spotting it!

ramonjd commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -281,6 +345,13 @@ export function PositionPanel( props ) {
 		web:
 			options.length > 1 ? (
 				<InspectorControls group="position">
+					{ firstParentClientId && value === 'sticky' ? (
```
[Ditto](https://github.com/WordPress/gutenberg/pull/49321/files#r1178521099)

ramonjd commented:
Working really well for me so far in both the editor and frontend:

https://user-images.githubusercontent.com/6458278/234734502-544884d7-e7a0-4ad3-843b-d47a8067d36e.mp4

Not sure if this is a problem or not, but I tested in the site editor and was missing the ability to set the header template to sticky. I had to create a new group block outside of the header template and set its position to sticky to get it to work. (Unless I'm missing something).

andrewserong commented:
Thanks for testing @ramonjd! 🙇 

> Not sure if this is a problem or not, but I tested in the site editor and was missing the ability to set the header template to sticky. I had to create a new group block outside of the header template and set its position to sticky to get it to work. (Unless I'm missing something).

Yep, that's still the case. Sticky support for template parts is a whole can of worms, unfortunately! There's an earlier PR that explored this (https://github.com/WordPress/gutenberg/pull/47133) with the conclusion being that adding support directly at the template part level creates a local change that doesn't propagate across all instances of the template part, which is a confusing experience for folks. Because of that, the template part block currently doesn't have any styling controls for it directly. It'd be a good thing to come up with a solution for, but beyond the scope of this PR 🙂

andrewserong committed:
Small code quality tweaks

jameskoster commented:
Sounds like it's more about communicating how sticky works generally, so I'd say context is essential for this to be intuitive. 

Applying a dotted outline to (or spotlighting) the parent when hovering / focussing the 'Sticky' option in the dropdown seems like a reasonable idea to try. 

Another option could be to apply the same behavior when hovering the block name in the help text. Crude mockup:

![parent](https://user-images.githubusercontent.com/846565/234826834-59037549-5d59-41c7-8cf7-e0e31f732110.gif)

andrewserong commented:
Thanks @jameskoster, that gives me good stuff to play with. If I don't get to it today, I'll have a play with that next week!

tellthemachines commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -222,32 +279,39 @@ export function PositionPanel( props ) {
 		[ clientId ]
 	);
 
+	const blockInformation = useBlockDisplayInformation( firstParentClientId );
```
Crazy idea: could we use this client id to instead inject a style in the DOM like `#block-clientid :has(.ist-sticky.is-selected) { outline: 1px solid grey}` or whatever?

andrewserong replied:
Oh, that is a clever idea — I can't remember why exactly, but I think from memory there was a preference to try to use popovers so that we don't accidentally collide with any styles that a theme might have for a particular block. It could be to ensure that the visualizers / popover always displays its CSS styling in addition to whatever might be set on the parent block (that we can't quite know) rather than overriding anything. Worth playing with though!

tellthemachines replied:
I picked outline because it's very unlikely that it gets used for block styling, as it's mostly a property applied on focus. Though if the popover calculations are performant I suppose it's not really an issue.

tellthemachines reviewed (commented):
I love the added flexibility of having this feature at a non-root level! Ideally we would also be able to apply it to non-Group blocks, but this is a great first step already 😄 

A few things I noticed when testing:

* If the sticky block is the only child of its parent, the outline is almost invisible. I anticipate this may be a common problem, as folks are bound to try and make sticky the Group block directly inside a Header template part:

<img width="888" alt="Screenshot 2023-04-28 at 2 25 06 pm" src="https://user-images.githubusercontent.com/8096000/235058555-cf238798-2ab6-4159-afa9-ae2f0771fabb.png">

* Any sticky elements in the template or post body will display over a mobile navigation overlay inside a sticky header. There's probably nothing we can do about this; it happens because a stacking context (created by `position:sticky`) that comes later in the DOM will overlay one that comes earlier. In most situations this would be desirable so I'm really not sure how best to approach it. 

<img width="390" alt="Screenshot 2023-04-28 at 2 39 42 pm" src="https://user-images.githubusercontent.com/8096000/235059445-5e62011a-9390-485e-a126-4668dbd36a23.png">

* The outline on the parent element behaves fine when resizing the browser window, but when opening/closing the list view it goes a bit bonkers:

<img width="1092" alt="Screenshot 2023-04-28 at 2 48 50 pm" src="https://user-images.githubusercontent.com/8096000/235059590-42944ee1-fd3a-4103-80a5-1399993ee6c9.png">

Also it seems to disappear altogether when settings sidebar is closed, though this might be a good thing?

andrewserong commented:
Thanks for the detailed testing @tellthemachines!
 
> If the sticky block is the only child of its parent, the outline is almost invisible. I anticipate this may be a common problem, as folks are bound to try and make sticky the Group block directly inside a Header template part:

Good point — we could always hide the visualizer if the parent only has one child so that there isn't a nearly overlapping line. Hopefully the text beneath the Sticky control will help (at least slightly) for folks that might be confused if they're adding it as a direct child. Another option could be to only display sticky as an option in non-root positions if the parent has more than one child... though that could be confusing for folks if the controls only appear some of the time 🤔

> Any sticky elements in the template or post body will display over a mobile navigation overlay inside a sticky header. There's probably nothing we can do about this; it happens because a stacking context (created by position:sticky) that comes later in the DOM will overlay one that comes earlier. In most situations this would be desirable so I'm really not sure how best to approach it.

Great catch! Yes, this is a similar issue to dealing with sticky headers in general, too, which I ran into the other day (https://github.com/WordPress/gutenberg/pull/49321#issuecomment-1522793257). I'm not too sure how best to approach it either, but I'll give it some more thought. At the moment a caveat of this PR is that the non-root sticky options don't play very nicely with sites that have sticky headers 😅. Happy to try any ideas folks might have!

> The outline on the parent element behaves fine when resizing the browser window, but when opening/closing the list view it goes a bit bonkers:

Another good catch! I think I might need to explore adding a resize observer somewhere that checks if the editor canvas has changed size, even though the window might not have. I'll dig in next week 👍

> Also it seems to disappear altogether when settings sidebar is closed, though this might be a good thing?

Yes, I think also from some of the other feedback it sounds like we might be reducing how frequently the visualizer is shown. I.e. if it's only shown while the sticky control is focused or hovered, then we don't need to worry about it hiding on closure.

Thanks again for the feedback everyone, I'll continue hacking on this next week!

tellthemachines commented:
> we could always hide the visualizer if the parent only has one child so that there isn't a nearly overlapping line

Might be best not to do this because we could well have a single child inside a column that is a lot taller than the child, in which case it would be useful to have the visualizer.

I was thinking more along the lines of making it _very_ visible, like bright red? so it can't be missed. This could work well especially if we're only showing it on hover/focus.

andrewserong commented:
> I was thinking more along the lines of making it very visible, like bright red? so it can't be missed. This could work well especially if we're only showing it on hover/focus.

Ah, gotcha, good thinking! Let's have a play with it and see what level of prominence feels just right. Thanks again for the great ideas!

andrewserong committed:
Try dotted outline, subtle display on focus and hover label

andrewserong commented:
Update: I've had a go at exploring some of James' ideas from this comment https://github.com/WordPress/gutenberg/pull/49321#issuecomment-1525310437 (and rebased this PR to include a fix for the Cover block and sticky position).

Here's how it's looking now using a subtle approach:

* Dotted outline
* Only displays when the dropdown is focused or the user hovers over the block name in the help text

https://user-images.githubusercontent.com/14988353/236392409-eb908116-8bf7-4bb8-a2d3-cef8dc106bdb.mp4

It's looking more elegant now, and I quite like that the border has a different appearance to the usual solid border to indicate that a block is selected. I do think it's possibly too subtle, though, and we might want it to be more prominent when the block is selected, rather than just when a user goes to change the value. And @tellthemachines raised good ideas about making it even more clear / prominent, so very happy to go in a different direction if it suits.

I'll be working on a few other things over the next week, so might be a little slower to update this particular PR, but happy for any ideas folks might have!

jameskoster commented:
Thanks for the update, this looks pretty good to me, though I'd appreciate more input from other @WordPress/gutenberg-design folks. 

The only detail I'd tweak is applying `cursor: default` to the block name in the help text.

jasmussen commented:
To compare with trunk, don't mind the sticky disappearing when choosing default, that's because it's not enabled for this block in trunk:

![before](https://user-images.githubusercontent.com/1204802/236416632-5f7ca6e8-8fc0-464f-bc1e-7aa603d5877a.gif)

Here's this branch:

![this branch](https://user-images.githubusercontent.com/1204802/236416984-09434b36-ffc2-421e-bf61-fa4dbee892d0.gif)

I appreciate the added help text, I think we can do thatm but I would use bog standard help text with no special underlines, and with the right metrics. So instead of this:

<img width="339" alt="Screenshot 2023-05-05 at 10 58 07" src="https://user-images.githubusercontent.com/1204802/236417157-f2b7792c-833f-4c47-ac0e-89fc78f8aa00.png">

I would do this:

<img width="332" alt="sticky" src="https://user-images.githubusercontent.com/1204802/236417527-6b5fefdb-19c1-428d-b9c6-41c22d186c1a.png">

Note the correct margin between help text and dropdown, and the lack of underline under "Columns". It's not a bad idea, but we don't use it anywhere else, and I think it reads well enough without it.

The more I'm thinking about this, however, the more anxious I am about adding new borders in the canvas. This dotted line does not feel substantial enough to really add context:

<img width="1392" alt="Screenshot 2023-05-05 at 10 58 14" src="https://user-images.githubusercontent.com/1204802/236418059-de7078e2-0a0c-4927-a560-dafff56ce701.png">

We also still have a task to try and eliminiate lines in the canvas, e.g. see https://github.com/WordPress/gutenberg/pull/44774 and https://github.com/WordPress/gutenberg/pull/44775, both which are stalled because we don't yet have a clear heuristic for when and how to show outlines. So I'm actually leaning towards us not wanting to add new outlines in the canvas until we have a clear ruleset for how to do this. Lest we risk add features just for this one block, when we should be thinking in more generic terms.

That said, I still hear the echoes for how this can be _visualized_. One of the things I think has been reasonably successful recently, is margin and padding visualizations:

![margin](https://user-images.githubusercontent.com/1204802/236418697-441b9ec9-18f1-4521-9d93-83ad4ddeb9a4.gif)

Note here the HUGE blue swathe of color that's instantly visible. Also note how it's _only_ ever shown when you hover or focus the control. I think we could use that same pattern here:

* Visualize the parent container using this blue color. It could be an overlay on the parent, in this case the Columns block. 
* Show this blue overlay _only_ when focusing or hovering the sticky control.

On a sidenote and unrelated to this PR, it looks like the focus style regressed:
<img width="142" alt="Screenshot 2023-05-05 at 11 09 24" src="https://user-images.githubusercontent.com/1204802/236419176-d2148709-9d65-4621-8374-f396b33a2b88.png">

<img width="303" alt="Screenshot 2023-05-05 at 11 09 33" src="https://user-images.githubusercontent.com/1204802/236419184-e0276a44-5727-4e4a-903e-5a0c4054b254.png">

Those should be 1.5px, not 2px. I'll look into it.

andrewserong committed:
Fix help text and spacing

andrewserong committed:
Update visualizer to only display on hover

ramonjd commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -214,40 +270,58 @@ export function PositionPanel( props ) {
+			? sprintf(
+					/* translators: %s: the name of the parent block. */
+					__(
+						'The block will stick to the scrollable area of the parent %s block.'
```
Not a blocker, just asking:

Is there a way to tell if a block is nested inside a template? 

I know [getting sticky to work on templates is off the cards for now](https://github.com/WordPress/gutenberg/pull/49321#issuecomment-1524407965), but I guess folks might try to make the header sticky and until we can get that working this doesn't make much sense when I try to make Header > Group sticky:

> The block will stick to the scrollable area of the parent Header block.

andrewserong replied:
> Is there a way to tell if a block is nested inside a template?

Good question!

Unfortunately it's a tricky one to figure out, because depending on what a user's site layout is like, there could be a good case for a user making their block sticky to the immediate parent within a header. I suppose it's one of the challenges of exposing the controls in non-root positions — we'd like to try to help communicate what's going to happen, but at the same time, it's up to users to use the tools however it suits their current designs 🤔

Personally, I'd probably explore tweaking logic surrounding where the block sits in relation to templates / template parts in follow-ups, as it gets quite tricky to come up with something the feels just right.

Thanks for taking a look!

andrewserong commented:
Update: I've given this a rebase and implemented the suggestions from @jasmussen.

* Fixed up spacing for the help text
* Removed hover on parent block link
* Updated visualizer styling so that it uses a background color instead of a border
* Updated visualizer display logic so it only shows when hovering over the select dropdown, and only when it's set to Sticky

Here's how it's looking:

![2023-05-26 14 29 32](https://github.com/WordPress/gutenberg/assets/14988353/ff4fffa0-37a8-4523-9183-a454e1b405c7)

For the original intention behind this PR, I think this is probably in pretty good shape for a final review now.

The main issue I'm looking into is arguably beyond the scope of this PR, but would be good to figure out, and that's the feedback from @tellthemachines in https://github.com/WordPress/gutenberg/pull/49321#pullrequestreview-1405225778 — in a site with a sticky header, if you set a block to be sticky within post content, then the one in post content will visually sit on top of both the sticky header, and the navigation block's overlay, due to the stacking context.

Here's a screengrab of the issue:

https://github.com/WordPress/gutenberg/assets/14988353/f1fc4c2a-4014-46c7-be27-dc61d97ffa43

There are two ways I can think of to resolve this:

* If a Group block is a wrapping block for a template part (i.e. the immediate parent) then give it a slightly higher z-index (say, `15` instead of `10`). This is quite easy to do in `position.php` as we have access to the block's inner blocks at render time. I haven't had much luck implementing this in JS as the current position CSS output doesn't appear to be able to access inner blocks at that stage. Happy to keep looking into it, though.
* Another alternative, and this could also be helpful for making sticky blocks that visually "stack" beneath a sticky header, would be to expose UI controls for the user to set z-index levels either directly or via a set of preset z-index values. By exploring this, we could also look at `top` position offsets, so that users can make sure their sticky blocks have enough clearance when going past the top of the screen.

Very happy for feedback on all this, and which level of features makes good enough sense for an incremental rollout. My feeling at the moment is that this PR is quite nice, but would be quite frustrating for users with sticky headers. The fix for automatically setting z-index to `15` could be good for that use case. Longer-term, I'd love to see user control over z-index and top position offsets, but this would likely require more design work to figure out a good UI / UX for it, and I imagine would need a bit more exploration so wouldn't be something we could do in time for 6.3.

Thanks again for the feedback and reviews, everyone! Those are my thoughts for the moment on the current state of this work for sticky positioning — I'll aim to continue chipping away at this from mid next week.

ramonjd commented:
> then the one in post content will visually sit on top of both the sticky header, and the navigation block's overlay, due to the stacking context.

For the mobile menu, would it be dodgy to hide `is-position-sticky` when the body has a class of `has-modal-open` ?

> If a Group block is a wrapping block for a template part (i.e. the immediate parent) then give it a slightly higher z-index (say, 15 instead of 10)

This sounds like a reasonable and non-invasive first approach for now. How long term would this be? I'm thinking backwards compatibility when and if we add z-index to block supports. I suppose `15` could become the fallback value for immediate children of templates.

Having said all that, I've tested this given those constraints and I think it's in good shape. Stickyness in most cases works as I expect and the visualizer appears when I hover over the stick item in the drop down only.

jasmussen commented:
This is cool, this could work:

![sticky](https://github.com/WordPress/gutenberg/assets/1204802/988d8ec4-7d01-4a0d-b63c-62b9bf916e30)

Can the blue indicator be _below_ the sticky block? I.e. indicating only the parent? Not crucial, but would be a nice little nod.

This one:

<img width="361" alt="Screenshot 2023-05-26 at 08 29 44" src="https://github.com/WordPress/gutenberg/assets/1204802/a206acbd-2a64-4239-af17-7cfda32a22a3">

The gray hover state for an item here is very dark and means the contrast isn't good. I _imagine_ it's a separate issue with the dropdown control, just like the 2px focus state, but mentioning here just in case it's local.

Would it be possible to also show the highlight if you hover the "Sticky" state in this dropdown? I.e. imagine it's set to default, and you hover the sticky state to show the parent container it'll stick to? Also not crucial at all, if it's hacky or even minimally hard to implement, not worth it, just wondering. 

I would love a check from @jameskoster but on my side, especially if we can put the highlight below the sticky block, I could get behind this!

andrewserong commented:
Thanks for taking this for another spin, Joen!

>  I imagine it's a separate issue with the dropdown control, just like the 2px focus state, but mentioning here just in case it's local.

Yes, I think that issue is with the control, but it'd be a good one to improve.

> Would it be possible to also show the highlight if you hover the "Sticky" state in this dropdown?

Unfortunately the `CustomSelectControl` doesn't let us pass event handlers down to each of the individual options. It'd be cool if it did, though — happy to look into extending that component either separately or in a follow-up so that it can allow events.

> Can the blue indicator be below the sticky block? I.e. indicating only the parent? Not crucial, but would be a nice little nod.

Oh, good idea. I'll have a play with that when I'm back next week. One potential issue is that if there isn't much content yet, then hovering over the dropdown might not show anything if we set the visualizer to be underneath the sticky block. So even if it doesn't look very pretty, it might still be useful for it to be overlaid so that it's clear when the sticky region is very minimal? Just a thought, but I don't feel too strongly about it.

The other challenge is that depending on how the popover works, it might be tricky to get it to display underneath 🤔... either way, I'll play with it and report back 🙂

jameskoster commented:
In terms of functionality this is working well, but the visualiser doesn't quite hit the mark for me. It may be subjective, but this treatment doesn't communicate the behavior in an intuitive way. Is it vital? Could we explore this detail separately?

andrewserong commented:
Thanks for re-testing!

>   but the visualiser doesn't quite hit the mark for me. It may be subjective, but this treatment doesn't communicate the behavior in an intuitive way. Is it vital? Could we explore this detail separately?

Unfortunately, I think having some form of visualisation is probably a requirement for reintroducing non-root sticky position blocks. Feedback when we initially introduced sticky position was that without having some form of visualisation, users unfamiliar with the CSS property would find the behaviour confusing that `sticky` means stick-to-parent, and I'm not sure if the help text is quite enough to communicate that on its own. I think @mtias and @richtabor might have been some of the folks who gave that feedback, from memory?

I'm happy to keep chipping away at this PR and/or try out other ideas. Also happy to merge in a reduced version of this PR that enables non-root sticky positioning + the help text, without the visualizer, if we can get consensus 🙂

mtias commented:
Can we try adding some visualization to the list view when this is set? (An icon with a tooltip, perhaps, or a badge that says "sticky".)

andrewserong commented:
> Can we try adding some visualization to the list view when this is set? (An icon with a tooltip, perhaps, or a badge that says "sticky".)

Yes, we started exploring this back in https://github.com/WordPress/gutenberg/pull/49122 but wound up being blocked by some work that's now landed to improve the accessibility of how blocks are read out in the list view. That work has landed now, so it should be ready for a rebase. I'll dust it off tomorrow.

ndiego commented:
Hey @andrewserong, this is looking great! Are we aiming for 6.3 with this? https://github.com/WordPress/gutenberg/issues/47043 is on the [6.3 Project Board](https://github.com/orgs/WordPress/projects/103/views/1), so I wasn't sure. If so, I'll add this PR to the board as well. Thanks!

andrewserong commented:
> Are we aiming for 6.3 with this? https://github.com/WordPress/gutenberg/issues/47043 is on the [6.3 Project Board](https://github.com/orgs/WordPress/projects/103/views/1), so I wasn't sure. If so, I'll add this PR to the board as well.

Thanks for the ping! Yes, my hope is to land both this and https://github.com/WordPress/gutenberg/pull/49122 as the main parts of the #47043 tracking issue for 6.3. It currently depends on whether we can reach a good balance in terms of design and functionality in time, but that is currently my aim.

ndiego commented:
Great, thanks @andrewserong. I updated the board accordingly.

andrewserong committed:
Remove visualizer

andrewserong commented:
Update: I have pared this PR back to just re-enabling non-root sticky positions, and the inclusion of the help text, and removed the visualizer for now. We also have https://github.com/WordPress/gutenberg/pull/49122 ready for review, which adds the sticky icon to the list view.

There is still an outstanding issue for sites that have sticky headers and also use sticky blocks elsewhere in the post content, as flagged by @tellthemachines earlier in https://github.com/WordPress/gutenberg/pull/49321#pullrequestreview-1405225778 — that issue technically exists irrespective of this PR, so I would lean toward looking into fixing that separately. I believe the main way to fix that would be to either allow individual control over z-index values, or add some clever logic that adjusts the z-index for the sticky block based on whether or not it contains a template part. Either way, I think it's best to look into fixes separately if we can.

I think this PR should be ready for a final review now. To recap, the PR now does the following:

* Allow non-root positions
* Add help text beneath the position control when Sticky is set and the block has a parent, to describe how the block will stick to its parent

Thanks again for the feedback, everyone! 🙇

ramonjd reviewed (approved):
Tested again in Chrome, Safari and FF

I think we're in a good spot and ready for further iterations. Thanks for "sticking" with this one!


https://github.com/WordPress/gutenberg/assets/6458278/e09f47c8-d6fc-47de-9770-b0e04f36417c

andrewserong commented:
Thanks for reviewing @ramonjd!

Since we have clearer ideas now about which visualisation aspects we do or don't want (no the visualiser, yes to adding an icon in the list view), and that the other changes in this PR appeared to be non-controversial (adding in help text, re-enabling non-root positions), I'll merge this one in now.

If there are any objections or issues over the weekend, I can spin up a revert on Monday — this PR only really changes whether or not the control is visible in non-root positions, so it'd be easy to return to the previous state, with plenty of time before the next Gutenberg RC.

Thanks again for all the input, everyone!
